File: merge_intervals.cpp
//leetcode 56 合并区间
//以数组 intervals 表示若干个区间的集合，其中单个区间为 intervals[i] = [start(i), end(i)] 。
//合并所有重叠的区间，并返回一个不重叠的区间数组，该数组需恰好覆盖输入中的所有区间。

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

vector<vector<int>> merge_intervals(vector<vector<int>> intervals) {
    if (intervals.empty()) return {};  // 检查是否为空
    sort(intervals.begin(), intervals.end());  // 先排序

    vector<vector<int>> res;
    for (const auto& interval : intervals) {
        int left = interval[0], right = interval[1];
        // 插入条件修改：当 res 为空或当前区间与 res 最后一个区间不重叠时
        if (res.empty() || res.back()[1] < left) {
            res.push_back({left, right});
        } else {
            // 更新当前的右边界
            res.back()[1] = max(res.back()[1], right);
        }
    }
    return res;
}

int main() {
    vector<vector<int>> intervals = {{1, 3}, {2, 6}, {8, 10}, {15, 18}};
    vector<vector<int>> res = merge_intervals(intervals);
    for (const auto& interval : res) {
        cout << "[" << interval[0] << ", " << interval[1] << "] ";
    }
    cout << endl;
    return 0;
}

File: MergeIntervals.java
import java.util.Arrays;
import java.util.List;
import java.util.LinkedList;;
public class MergeIntervals {
    public int[][] merge(int[][] intervals){
        if(intervals.length<=1)return intervals;
        Arrays.sort(intervals,(a,b)->a[0]-b[0]);
        List<int[]>res=new LinkedList<>();
        for(int[] interval:intervals){
            if(res.isEmpty()&&res.getLast()[1]<interval[0]){
                res.add(interval);
            }
            else{
                res.getLast()[1]=Math.max(res.getLast()[1],interval[1]);
            }
        }
        return res.toArray(new int[res.size()][]);
    }

    public static void main(String args){
        MergeIntervals mergeintervals=new MergeIntervals();
        int[][] intervals = {
            {1, 3},
            {2, 6},
            {8, 10},
            {15, 18}
        };
        int[][]ans=mergeintervals.merge(intervals);
        System.out.println("合并后的区间：");
        for (int[] an :ans ) {
            System.out.println("[" + an[0] + ", " + an[1] + "]");
        }
    }
}